ac-34 product transactions api doc

ExtraGravity opened the pull request.

ExtraGravity committed:
ac-34 product transactions api doc

beaesguerra commented on src/routers/transactions/products.ts:
```diff
@@ -8,14 +8,96 @@ import { StatusCode } from '@/models/statusCodes';
+ *                - costPerUnit
+ *                - currency
+ *              Parameters have no effect on this request.
+ *              Associated attributes can be checked via /products/<type>
```
isn't it tedious to have to make 2 requests to be able to use product transaction data? for example, you'd make this kind of request, but in order for a client to present it properly, it will also need to make a request for /products/<type> to know what unit is being used. could we just do the join in the core so it's available in this response?

beaesguerra replied:
also in the docs, it'd be very helpful if you gave an example of what an associated attribute is (ie. density for milk product transactions)

ExtraGravity replied:
This does do the join as well. You can see in the example it's there. I added the line 

```
 *              An example of an associated attribute would be density as
 *              in the example.
```

To make it clearer.

beaesguerra commented on src/routers/transactions/products.ts:
```diff
@@ -8,14 +8,96 @@ import { StatusCode } from '@/models/statusCodes';
+ *              Associated attributes can be checked via /products/<type>
+ * @apiName GetAllProductTypeTransactions
+ *
+ * @apiError (401) Unauthorized - Must be admin
```
should be 403

beaesguerra commented on src/routers/transactions/products.ts:
```diff
@@ -8,14 +8,96 @@ import { StatusCode } from '@/models/statusCodes';
+  [
+    {
+      "productTransactionUuid": "3cad5e7c-5444-4de1-aa81-a7d15acb35f1",
+      "datetime": "2018-01-23 04:05:06",
```
doesn't utc have a specific format? doesn't this need a z at the end? that way clients will know that the datetime they are receiving is for sure UTC

beaesguerra replied:
@Right2Drive is requesting that we add milliseconds to the datetime as well

ExtraGravity replied:
I feel like milliseconds is not useful for this application. But we can add a timezone, it should always be UTC.

beaesguerra commented on src/routers/transactions/products.ts:
```diff
@@ -8,14 +8,96 @@ import { StatusCode } from '@/models/statusCodes';
+ * @api {post} /transactions/products/<type>
+ * @description Returns all product transactions of a specified type.
+ *              Required fields are:
+ *                - datetime
```
need @apiparm tags for these

beaesguerra replied:
how have we been specifying how to pass in parameters in the request? query parameters or request body params?

ExtraGravity replied:
It should be implied it's a body param since you're supposed to specify in the URL at the top for any URL or Query params.

Which I have not been doing properly :) 

Instead of 
```
@api {post} /transactions/products/<type>
```
it should be
```
@api {post} /transactions/products/:type
```
And include it as an API param.

beaesguerra commented on src/routers/transactions/products.ts:
```diff
@@ -8,14 +8,96 @@ import { StatusCode } from '@/models/statusCodes';
+ *                - costPerUnit
+ *                - currency
+ *                Along with any associated attributes.
+ *                Associated attirubtes can be checked via /products
```
typo. maybe add example here of an associated attribute?

beaesguerra commented on src/routers/transactions/products.ts:
```diff
@@ -8,14 +8,96 @@ import { StatusCode } from '@/models/statusCodes';
+ * @apiName CreateNewProductTransaction
+ *
+ * @apiError (400) The required fields are invalid or missing, ...
+ * @apiError (401) Unauthorized - Must be admin
```
403 forbidden here

beaesguerra commented on src/routers/transactions/products.ts:
```diff
@@ -8,14 +8,96 @@ import { StatusCode } from '@/models/statusCodes';
+ *
+ * @apiName CreateNewProductTransaction
+ *
+ * @apiError (400) The required fields are invalid or missing, ...
```
do we need to specify it's a bad request? this may get formatted funny

beaesguerra commented on src/routers/transactions/products.ts:
```diff
@@ -8,14 +8,96 @@ import { StatusCode } from '@/models/statusCodes';
+ * @apiError (400) The required fields are invalid or missing, ...
+ * @apiError (401) Unauthorized - Must be admin
+ *
+ * @apiSuccess (201) {String} Successfully created new <type> transaction
```
haha you'll have to do the proper formatting again to make it generate nicely

ExtraGravity replied:
This is not being a supportive team mate

beaesguerra commented on src/routers/transactions/products.ts:
```diff
@@ -8,14 +8,96 @@ import { StatusCode } from '@/models/statusCodes';
+/**
+ * @api {put} /transactions/products/<type>/<uuid>
+ * @description Updates a product transaction.
+ *              Available fields are:
```
specify that they are not all mandatory? or are we making them all mandatory?

beaesguerra replied:
need apiParam tag here :)

beaesguerra commented on src/routers/transactions/products.ts:
```diff
@@ -8,14 +8,96 @@ import { StatusCode } from '@/models/statusCodes';
+ *
+ * @apiName UpdateProductTransaction
+ *
+ * @apiError (400) The fields are invalid, ...
```
need bad request title?

beaesguerra commented on src/routers/transactions/products.ts:
```diff
@@ -8,14 +8,96 @@ import { StatusCode } from '@/models/statusCodes';
+ * @apiName UpdateProductTransaction
+ *
+ * @apiError (400) The fields are invalid, ...
+ * @apiError (401) Unauthorized - Must be admin
```
forbidden 403

ExtraGravity committed:
Merge branch 'master' into enoch/ac-34-product-transactions-api-doc

ExtraGravity committed:
ac-34 follow proper apidoc syntax in product transactions api

ExtraGravity committed:
Add last modified date to product transactions, and information on error update

Clients use this information for syncing so it should be as precise as possible.
Clients also write to internal store before sending an update, so returning correct
information on a failed write is a useful utility.

beaesguerra commented on src/routers/transactions/products.ts:
```diff
@@ -8,14 +8,117 @@ import { StatusCode } from '@/models/statusCodes';
+ *
+ * @apiParam {String} type The type of product.
+ * @apiParam {String} datetime The time this transaction was conducted.
+ *                    It value must be in UTC or the request will be rejected.
```
should be It's.

beaesguerra commented on src/routers/transactions/products.ts:
```diff
@@ -8,14 +8,117 @@ import { StatusCode } from '@/models/statusCodes';
+ * @apiVersion  0.0.1
+ * @description Updates a product transaction.
+ *              Available fields are:
+ *                - datetime
```
we shouldn't be able to update the datetime?

beaesguerra commented on src/routers/transactions/products.ts:
```diff
@@ -8,14 +8,117 @@ import { StatusCode } from '@/models/statusCodes';
+ *                - datetime
+ *                - toPersonUuid
+ *                - fromPersonUuid
+ *                - productTypeId
```
this list doesn't need to be here since we are using apiparam

beaesguerra commented on src/routers/transactions/products.ts:
```diff
@@ -38,6 +39,7 @@ const router = createRouter();
   [
     {
       "productTransactionUuid": "3cad5e7c-5444-4de1-aa81-a7d15acb35f1",
+      "lastModified": "2018-01-23 04:05:06"
```
need the Z?

beaesguerra replied:
add milliseconds 

2018-01-23 04:05:06.789Z

beaesguerra commented on src/routers/transactions/products.ts:
```diff
@@ -105,16 +107,18 @@ router.post('/:type', async (req, res) => {
  *                - currency
  *                Along with any associated attributes such as density.
  *                Associated attributes can be checked via /products
+ *                Returns the transaction's actual data on error.
```
actual data as in the request data?

beaesguerra commented on src/routers/transactions/products.ts:
```diff
@@ -105,16 +107,18 @@ router.post('/:type', async (req, res) => {
  *                    All attributes must be provided in separate params.
  *
  * @apiError (400) BadRequest The following fields are invalid, ...
+ *                 The product's actual data is returned in JSON format.
```
does this mean the request data?

Commit:
ac-50 address code review comments

Commit:
fixing comments
